service/tbc/zkindexer.go

```diff
@@ -0,0 +1,235 @@
+	"github.com/btcsuite/btcd/txscript"
+
+	"github.com/hemilabs/heminetwork/v2/database"
+	"github.com/hemilabs/heminetwork/v2/database/tbcd"
+)
+
+// Transactions to be looked up by txid.
+
+type zkIndexer struct {
+	indexerCommon
+
+	cacheCapacity int
+}
+
+var (
+	_ Indexer = (*zkIndexer)(nil)
+	_ indexer = (*zkIndexer)(nil)
+)
+
+func NewZKIndexer(g geometryParams, cacheLen int, enabled bool) Indexer {
+	zi := &zkIndexer{
+		cacheCapacity: cacheLen,
+	}
+	zi.indexerCommon = indexerCommon{
+		name:    "zkindexer",
+		enabled: enabled,
+		g:       g,
+		p:       zi,
+	}
+	return zi
+}
+
+func (i *zkIndexer) newCache() indexerCache {
+	return NewCache[tbcd.ZKIndexKey, []byte](i.cacheCapacity)
+}
+
+func (i *zkIndexer) indexerAt(ctx context.Context) (*tbcd.BlockHeader, error) {
+	bh, err := i.g.db.BlockHeaderByZKIndex(ctx)
+	return i.evaluateBlockHeaderIndex(bh, err)
+}
+
+func (i *zkIndexer) balance(ctx context.Context, ss tbcd.ScriptHash, c indexerCache) ([]byte, error) {
+	cache := c.(*Cache[tbcd.ZKIndexKey, []byte]).Map()
+	if balance, ok := cache[tbcd.ZKIndexKey(ss[:])]; ok {
+		return balance, nil
+	}
+
+	// Not found in cache, fetch from db
+	var balance [8]byte
+	b, err := i.g.db.ZKBalanceByScriptHash(ctx, ss)
+	if err != nil {
+		if errors.Is(err, database.ErrNotFound) {
+			return balance[:], nil
+		}
+		return nil, err
+	}
+	binary.BigEndian.PutUint64(balance[:], b)
+	return balance[:], nil
+}
+
+func (i *zkIndexer) txOut(ctx context.Context, pop tbcd.Outpoint, c indexerCache) (uint64, []byte, error) {
+	cache := c.(*Cache[tbcd.ZKIndexKey, []byte]).Map()
+	if utxoE, ok := cache[tbcd.ZKIndexKey(pop[:])]; ok {
+		txOut := tbcd.TxOutFromBytes(utxoE)
+		return uint64(txOut.Value), txOut.PkScript, nil
+	}
+	// Not found in cache, fetch from db
+	value, txOut, err := i.g.db.ZKValueAndScriptByOutpoint(ctx, pop)
+	if err != nil {
+		return 0, nil, err
+	}
+	return value, txOut, nil
+}
+
+func (i *zkIndexer) processTx(ctx context.Context, direction int, blockHeight uint32, blockHash *chainhash.Hash, tx *btcutil.Tx, c indexerCache) error {
+	cache := c.(*Cache[tbcd.ZKIndexKey, []byte]).Map()
+	txId := tx.Hash()
+	for txInIdx, txIn := range tx.MsgTx().TxIn {
+		// Skip coinbase inputs
+		if blockchain.IsCoinBase(tx) {
+			continue
+		}
+
+		// Recreate Outpoint from TxIn.PreviousOutPoint
+		pop := tbcd.NewOutpoint(txIn.PreviousOutPoint.Hash,
+			txIn.PreviousOutPoint.Index)
+
+		// Retrieve TxOut from PreviousOutPoint
+		value, script, err := i.txOut(ctx, pop, c)
+		if err != nil {
+			return fmt.Errorf("tx out: %w", err)
+		}
+
+		// Retrieve balance
+		sh := tbcd.NewScriptHashFromScript(script)
+		balance, err := i.balance(ctx, sh, c)
+		if err != nil {
+			return fmt.Errorf("balance in: %w", err)
+		}
+
+		// Handle balance
+		switch direction {
+		case 1:
+			cache[tbcd.ZKIndexKey(sh[:])] = tbcd.BESubUint64(balance,
+				value)
+		case -1:
+			cache[tbcd.ZKIndexKey(sh[:])] = tbcd.BEAddUint64(balance,
+				value)
+		default:
+			panic("wtf")
```

Comment:
We should be using actual error messages for these:
```suggestion
			panic(fmt.Errorf("bug: invalid direction: %v", direction))
```

***Same for other `panic("wtf")` calls***